[REFACTOR] Avoid computing samples eagerly in SimulationResults

mattdailis opened the pull request:
* **Tickets addressed:** N/A
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
<!-- What approach was taken to satisfy the ticket being addressed? What should reviewers be aware of? -->
This PR removes `resourceSamples` as a field on `SimulationResults`. Prior to this PR, this field was derived from `realProfiles` and `discreteProfiles` via the `takeSamples` method - upon calling the `SimulationResults` constructor, `takeSamples` would be called eagerly.

In typical usage, `resourceSamples` is unused - the simulation results are written to the database using the `realProfiles` and `discreteProfiles` directly. The only time it is used (outside of testing) is via the explicit `resourceSamples` hasura action.

This PR moves computation of resource samples out of the `SimulationResults` constructor, so that they can be computed when requested, but not by default.

I have not done any performance measurements to see whether or not this change is significant - nevertheless, I believe avoiding needless computations is generally a good thing.

## Verification
<!-- How were the changes validated? Were any automated tests added, updated, removed, or re-baselined? -->
I spun up a local Aerie deployment and checked that running simulations and viewing their results worked, and I also called the `resourceSamples` endpoint directly.

## Documentation
<!-- What documentation was invalidated by these changes? Which artifacts should reviewers check for accuracy and completeness? -->
No functional behavior has changed, so no documentation needs to be updated.

## Future work
<!-- What next steps can we anticipate from here, if any? -->
No future work

mattdailis committed:
Avoid computing samples eagerly in SimulationResults

There are now two separate flows - when a simulation is run, and when
samples are requested explicitly. Samples are not written to the
database, so there is no need to compute samples unless they are
requested.

Mythicaeda commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/GetSimulationResultsAction.java:
```diff
@@ -78,7 +77,7 @@ public Map<String, List<Pair<Duration, SerializedValue>>> getResourceSamples(fin
     return this.simulationService
         .get(planId, revisionData)
-        .map(r -> r.resourceSamples)
+        .map(SimulationResults::resourceSamples)
```
Could you clarify where `resourceSamples` gets the `SimulationResults` parameter it needs here?

mattdailis replied:
Sure thing - `map` provides the simulation results as an argument - the code below is equivalent: 

```suggestion
        .map(results -> SimulationResults.resourceSamples(results))
```

Mythicaeda replied:
Thanks!

Mythicaeda reviewed (commented):
Looks fine, although seeing it appear in the mission models makes me wonder how often this calculation will be called per SimulationResults object. Would it be worth it to instead make `resourceSamples` a private field that is not assigned at construction, and have it accessed through a getter that assigns the output of `takeSamples` into it if `resourceSamples` is null? Then we only run `takeSamples` at most once per SimulationResults object.

mattdailis commented:
> Looks fine, although seeing it appear in the mission models makes me wonder how often this calculation will be called per SimulationResults object. Would it be worth it to instead make `resourceSamples` a private field that is not assigned at construction, and have it accessed through a getter that assigns the output of `takeSamples` into it if `resourceSamples` is null? Then we only run `takeSamples` at most once per SimulationResults object.

If we expected this computation to happen multiple times, absolutely, I think caching it in a field would be a valid approach. 💯 

As it stands, we've moved away from this method (it provides strictly less information than `realProfiles` and `discreteProfiles` do, while using more memory) - so we don't expect this method to get more use in the future. If I had my druthers, we'd remove this method entirely, but we're keeping the `resourceSamples` hasura action in the API as a convenience for clients that don't want to work with continuous profiles.

Mythicaeda commented:
> As it stands, we've moved away from this method (it provides strictly less information than `realProfiles` and `discreteProfiles` do, while using more memory) - so we don't expect this method to get more use in the future. If I had my druthers, we'd remove this method entirely, but we're keeping the `resourceSamples` hasura action in the API as a convenience for clients that don't want to work with continuous profiles.

So this refactor is setup for an eventual deprecation? If so, then I think I'm fine with making it the responsibility of whoever calls the public method to cache its return value.

mattdailis commented:
> So this refactor is setup for an eventual deprecation? If so, then I think I'm fine with making it the responsibility of whoever calls the public method to cache its return value.

When I brought up the possibility of deprecation, @patkenneally explained that we should keep the capability in case API clients want it, but that we should encourage people to use the continuous profiles directly. So - yes, it's "discouraged" if not exactly "deprecated for removal".

Twisol commented:
Personally, I'd like to see the `resourceSamples` code moved out of `SimulationResults` entirely. It looks like the only place where we access it outside of testing/development code is that API endpoint, and the computation doesn't really feel like the responsibility of `SimulationResults` anymore.

mattdailis committed:
Move takeSamples out of SimulationResults and into GetSimulationResultsAction::getResourceSamples

mattdailis commented:
> Personally, I'd like to see the resourceSamples code moved out of SimulationResults entirely

Done :+1: The sampling logic is now the responsibility of the `GetSimulationResultsAction`, which was the only client of this functionality. I inlined it into the `getResourceSamples` method.

mattdailis committed:
Add missing parameter to plan.json

This plan.json is used in the main method in SimulateMapSchedule -
it's not used for any automated tests, but it can be used for manual
testing. In #334, we added a z parameter to the FooActivity, but
forgot to update this file.

Twisol commented:
Awesome! Thanks for doing that!

Mythicaeda commented on examples/foo-missionmodel/src/test/resources/gov/nasa/jpl/aerie/foomissionmodel/plan.json:
```diff
@@ -2,24 +2,29 @@
       { "x": 1,
-        "y": "test_1" } },
+        "y": "test_1",
+        "z": 4 } },
```
Quick sanity check: where did the new `z` argument come from?

mattdailis replied:
Good question :+1: It came from #334, when we wanted to include a "required" parameter in our sample mission models - since we had a regression where required parameters would fail to instantiate.

This plan.json is used in the main method in `SimulateMapSchedule` - it's not used for any automated tests, but it can be used for manual testing. We should have added the z parameter in #334. 

It's unrelated to this PR - so I would accept the argument that it should be split into its own PR :+1: But it is such a small, inconsequential change, that I'd also be okay with bundling it into this PR.

Twisol replied:
It's in its own commit, so I think it's fine. IMHO, we should feel comfortable cleaning up things as we come in contact with code, even if it's technically not required for the task at hand.

Twisol replied:
It would be / would have been good to include that context in the commit message. (Remember, the title needs to be short, but you can have as long a commit body as you desire.)
